File: back/src/main/java/com/assettagging/spotit/system/business/service/SystemService.java
package com.assettagging.spotit.system.business.service;


import java.util.List;
import java.util.Map;

import com.assettagging.spotit.system.domain.model.DexConfiguration;
import com.assettagging.spotit.system.domain.model.DexModule;
import com.assettagging.spotit.system.domain.model.DexReferenceNo;
import com.assettagging.spotit.system.domain.model.DexSubModule;
import com.assettagging.spotit.system.domain.model.*;

/**
 */
public interface SystemService {

    //==============================================================================================
    // MODULE SUB MODULE
    //==============================================================================================

    DexModule findModuleById(Long id);

    DexSubModule findSubModuleById(Long id);

    DexModule findModuleByCode(String code);

    DexSubModule findSubModuleByCode(String code);

    List<DexModule> findModules();

    List<DexModule> findModules(boolean authorized);

    List<DexModule> findModules(Integer offset, Integer limit);

    List<DexModule> findModules(boolean authorized, Integer offset, Integer limit);

    List<DexSubModule> findSubModules();

    List<DexSubModule> findSubModules(boolean authorized);

    List<DexSubModule> findSubModules(Integer offset, Integer limit);

    List<DexSubModule> findSubModules(DexModule module, Integer offset, Integer limit);

    Integer countModule();

    Integer countSubModule();

    Integer countSubModule(DexModule module);

    void saveModule(DexModule module);

    void updateModule(DexModule module);

    void addSubModule(DexModule module, DexSubModule subModule);

    void updateSubModule(DexModule module, DexSubModule subModule);

    //==============================================================================================
    // REFERENCE NO
    //==============================================================================================

    DexReferenceNo findReferenceNoById(Long id);

    DexReferenceNo findReferenceNoByCode(String code);

    List<DexReferenceNo> findReferenceNos(Integer offset, Integer limit);

    List<DexReferenceNo> findReferenceNos(String filter, Integer offset, Integer limit);

    Integer countReferenceNo();

    Integer countReferenceNo(String filter);

    void saveReferenceNo(DexReferenceNo referenceNo);

    void updateReferenceNo(DexReferenceNo referenceNo);

    void removeReferenceNo(DexReferenceNo referenceNo);

    String generateReferenceNo(String code);

    String generateFormattedReferenceNo(String code, Map<String, Object> map);

    //==============================================================================================
    // CONFIGURATION
    //==============================================================================================

    DexConfiguration findConfigurationById(Long id);

    DexConfiguration findConfigurationByKey(String key);

    List<DexConfiguration> findConfigurationsByPrefix(String prefix);

    List<DexConfiguration> findConfigurations();

    List<DexConfiguration> findConfigurations(Integer offset, Integer limit);

    List<DexConfiguration> findConfigurations(String filter);

    List<DexConfiguration> findConfigurations(String filter, Integer offset, Integer limit);

    Integer countConfiguration();

    Integer countConfiguration(String filter);

    void saveConfiguration(DexConfiguration config);

    void updateConfiguration(DexConfiguration config);

    void removeConfiguration(DexConfiguration config);
}

File: front/src/app/services/system.service.ts
import {Observable} from 'rxjs';
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {environment} from '../../environments/environment';
import {ReferenceNo} from '../modules/system/reference-nos/reference-no.model';
import {Configuration} from '../modules/system/configurations/configuration.model';
import {ConfigurationResult} from '../modules/common/configurations/configuration.model';

@Injectable()
export class SystemService {

  private SYSTEM_API: string = environment.endpoint + '/api/system';

  constructor(private http: HttpClient) {
  }

  // ===================================================================================================================
  // REFERENCE NO
  // ===================================================================================================================

  findReferenceNos(): Observable<ReferenceNo[]> {
    return this.http.get<ReferenceNo[]>(this.SYSTEM_API + '/reference-nos');
  }

  findReferenceNoByCode(code: string): Observable<ReferenceNo> {
    return this.http.get<ReferenceNo>(this.SYSTEM_API + '/reference-nos/' + code);
  }

  saveReferenceNo(code: ReferenceNo) {
    return this.http.post(this.SYSTEM_API + '/reference-nos', JSON.stringify(code));
  }

  updateReferenceNo(code: ReferenceNo) {
    return this.http.put(this.SYSTEM_API + '/reference-nos/' + code.code, JSON.stringify(code));
  }

  removeReferenceNo(code: ReferenceNo) {
    return this.http.delete(this.SYSTEM_API + '/reference-nos/' + code.code);
  }

  // ===================================================================================================================
  // REFERENCE NO
  // ===================================================================================================================

    findPagedConfigurations(filter: string, page: number): Observable<ConfigurationResult> {
        const url = `${this.SYSTEM_API}/configurations?filter=${filter}&page=${page}`;
        return this.http.get<ConfigurationResult>(url);
    }

  findConfigurations(): Observable<Configuration[]> {
    return this.http.get<Configuration[]>(this.SYSTEM_API + '/configurations');
  }

  findConfigurationByKey(key: string): Observable<Configuration> {
    return this.http.get<Configuration>(this.SYSTEM_API + '/configurations/' + key);
  }

  saveConfiguration(key: Configuration) {
    return this.http.post(this.SYSTEM_API + '/configurations', JSON.stringify(key));
  }

  updateConfiguration(key: Configuration) {
    return this.http.put(this.SYSTEM_API + '/configurations/' + key.key, JSON.stringify(key));
  }

  removeConfiguration(key: Configuration) {
    return this.http.delete(this.SYSTEM_API + '/configurations/' + key.key);
  }
}
